go/ocr2/decryptionplugin/decryption.go

```diff
@@ -0,0 +1,375 @@
+	for _, request := range decryptionRequests {
+		ciphertext := &tdh2easy.Ciphertext{}
+		err := ciphertext.UnmarshalVerify(request.ciphertext, dp.publicKey)
+		if err != nil {
+			dp.logger.Error("DecryptionReporting Observation: cannot unmarshal the ciphertext, skipping it", commontypes.LogFields{
+				"error":        err,
+				"ciphertextID": request.ciphertextId,
+			})
+			continue
+		}
+		queryProto.DecryptionRequests = append(queryProto.GetDecryptionRequests(), &CiphertextWithID{
+			CiphertextID: request.ciphertextId,
+			Ciphertext:   request.ciphertext,
+		})
+	}
+
+	dp.logger.Debug("DecryptionReporting Query: end", commontypes.LogFields{
+		"epoch":    ts.Epoch,
+		"round":    ts.Round,
+		"queryLen": len(queryProto.DecryptionRequests),
+	})
+	queryProtoBytes, err := proto.Marshal(&queryProto)
+	if err != nil {
+		return nil, fmt.Errorf("cannot marshal query: %w", err)
+	}
+	return queryProtoBytes, nil
+}
+
+// Observation creates a decryption share for each request in the query.
+// If dp.specificConfig.Config.LocalRequest is true, then the oracle
+// only creates a decryption share for the decryption requests which it has locally.
+func (dp *decryptionPlugin) Observation(ctx context.Context, ts types.ReportTimestamp, query types.Query) (types.Observation, error) {
+	dp.logger.Debug("DecryptionReporting Observation: start", commontypes.LogFields{
+		"epoch": ts.Epoch,
+		"round": ts.Round,
+	})
+
+	queryProto := &Query{}
+	if err := proto.Unmarshal(query, queryProto); err != nil {
+		return nil, fmt.Errorf("cannot unmarshal query: %w", err)
+	}
+
+	observationProto := Observation{}
+	for _, request := range queryProto.DecryptionRequests {
+		ciphertext := &tdh2easy.Ciphertext{}
+		ciphertextBytes := request.Ciphertext
+		if err := ciphertext.UnmarshalVerify(ciphertextBytes, dp.publicKey); err != nil {
+			dp.logger.Error("DecryptionReporting Observation: cannot unmarshal and verify the ciphertext, the leader is faulty", commontypes.LogFields{
+				"error":        err,
+				"ciphertextID": request.CiphertextID,
+			})
+			return nil, fmt.Errorf("cannot unmarshal and verify the ciphertext: %w", err)
+		}
+		if dp.specificConfig.Config.LocalRequest {
+			queueCiphertextBytes, err := dp.decryptionQueue.GetCiphertext(request.CiphertextID)
+			if err != nil {
+				dp.logger.Error("DecryptionReporting Observation: cannot find ciphertext locally, skipping it", commontypes.LogFields{
+					"ciphertextID": request.CiphertextID,
+				})
+				continue
+			}
+			if !bytes.Equal(queueCiphertextBytes, ciphertextBytes) {
+				dp.logger.Error("DecryptionReporting Observation: local ciphertext does not match the query ciphertext, skipping it", commontypes.LogFields{
+					"ciphertextID": request.CiphertextID,
+				})
+				continue
+			}
+		}
+
+		decryptionShare, err := ciphertext.Decrypt(dp.privKeyShare)
+		if err != nil {
+			dp.logger.Error("DecryptionReporting Observation: cannot decrypt the ciphertext, the leader is faulty", commontypes.LogFields{
```

Comment:
Let's remove "the leader is faulty" part. There could be other reasons for this failure.